p11-kit/rpc-client.c

```diff
@@ -582,7 +582,7 @@ proto_read_sesssion_info (p11_rpc_message *msg,
 #define IN_ULONG_BUFFER(arr, len) \
 	if (len == NULL) \
 		{ _ret = CKR_ARGUMENTS_BAD; goto _cleanup; } \
-	if (!p11_rpc_message_write_ulong_buffer (&_msg, arr ? *len : 0)) \
+	if (!p11_rpc_message_write_ulong_buffer (&_msg, arr ? sizeof(*arr): 0, *len)) \
```

Comment:
Here the second argument of `p11_rpc_message_write_ulong_buffer` is a flag that indicates whether `arr` is NULL or not, right?

In that case, how about changing the prototype of `p11_rpc_message_write_ulong_buffer` to:
```c
bool
p11_rpc_message_write_ulong_buffer (p11_rpc_message *msg,
                                    CK_ULONG_PTR array,
                                    CK_ULONG_PTR count)
```
and directly passing the arguments?

Reply:
thanks for your review.

I cannot: if I do so, the sizeof(*arr) will get lost. CK_ULONG_PTR does not provide any information about array itself.

The beauty of IN_ULONG_BUFFER() is to be within the context of the callers so we still have the size within this context.

For instance, how to carry the size of CK_SLOT_ID ?

Reply:
There seems to be some misunderstanding. First, `sizeof(*arr)` doesn't return the number of elements, but the pointer size, which is always 4 or 8.  Second, I don't think we need to carry the array length in the first place.

Here is a commit I meant:
https://github.com/ueno/p11-kit/commit/33212be41344579aa517e46dda0067c2368c956a

Reply:
I was assuming that the typeof *arr could be anything, so the size could be anything beside 4 or 8 bytes (it could be any kind of structure).

If we assume that *arr is not "anything" then yes, I do agree with you. Then, I agree that it makes the API clearer so we avoid to carry both a length and a count (another length).

Do you want me to update all the serie as you suggest ?

I have seen that you have updated some unit tests too: I could not find how to run these tests. Do you have a pointer/URL/documentation or some notes I could follow in order to check the tests on a Ubuntu environment ?

Reply:
> I was assuming that the typeof *arr could be anything, so the size could be anything beside 4 or 8 bytes (it could be any kind of structure).

That's not how C works, I'm afraid :-)

> Do you want me to update all the serie as you suggest ?

Sure.

> I have seen that you have updated some unit tests too: I could not find how to run these tests. Do you have a pointer/URL/documentation or some notes I could follow in order to check the tests on a Ubuntu environment ?

If you use meson, do:
```sh
$ meson _build
$ ninja -C _build
$ ninja -C _build test
```
https://mesonbuild.com/Running-Meson.html#running-tests

If you use autotools, do:
```sh
$ ./configure ...
$ make
$ make check
```
https://www.gnu.org/software/automake/manual/html_node/Tests.html#Tests

You can also look at the logs on travis, which runs all tests.

Reply:
thanks, done based on your advises:

> TOTAL: 759
> PASS:  759